### PR TITLE
[DLCov 1/5] Add CMake option to enable enhanced line number coverage tracking

### PR DESCRIPTION
**Pull request by SLTozer**
This is part of a series of patches that tries to improve DILocation bug detection in Debugify; see below for more details. This first patch adds the necessary CMake flag to LLVM and a variable defined by that flag to LLVM's config header, allowing the next patch to track information without affecting normal builds.

--------------------

This series of patches adds a "DebugLoc coverage tracking" feature, that inserts conditionally-compiled tracking information into DebugLocs (and by extension, to Instructions), which is used by Debugify to provide more accurate and detailed coverage reports. When enabled, this features tracks whether and why we have intentionally dropped a DebugLoc, allowing Debugify to ignore false positives. An optional additional feature allows also storing a stack trace of the point where a DebugLoc was unintentionally dropped/not generated, which is used to make fixing detected errors significantly easier. The goal of these features is to provide useful tools for developers to fix existing DebugLoc errors and allow reliable detection of regressions by either manual inspection or an automated script.

Next: https://github.com/llvm/llvm-project/pull/107279

### CODE REVIEW

**Comment by djtodoro** [comment]
@SLTozer Thanks a lot for this! Very interesting.

I will review the implementation, but I am curious whether we only modified the metadata/`DebugLoc`. To be more precise, my main concern is memory usage increase, and whether we have affected `llvm::Instruction` in any way other than metadata.

**Comment by SLTozer** [comment]
> I will review the implementation, but I am curious whether we only modified the metadata/`DebugLoc`. To be more precise, my main concern is memory usage increase, and whether we have affected `llvm::Instruction` in any way other than metadata.

There will be increased memory usage - more exactly, we increase the size of `DebugLoc`, _not_ `!DILocation`, which means the extra storage is part of the `Instruction` class. The intent here is that we conditionally compile this extra information in for a specific build configuration designed to test the compiler itself with debugify; in ordinary builds, including debug builds, the extra information is not stored, so run time and memory usage should be unaffected.

File: llvm/CMakeLists.txt [hunk 1]
```diff
@@ -524,6 +524,10 @@ endif()
 
 option(LLVM_ENABLE_CRASH_DUMPS "Turn on memory dumps on crashes. Currently only implemented on Windows." OFF)
 
+set(LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING "DISABLED" CACHE STRING
+  "Enhance debugify's line number coverage tracking; enabling this is abi-breaking. Can be DISABLED, COVERAGE, or COVERAGE_AND_ORIGIN.")
```
Comment (jryans) [inline llvm/CMakeLists.txt:528]:
```suggestion
  "Enhance Debugify's line number coverage tracking; enabling this is ABI-breaking. Can be DISABLED, COVERAGE, or COVERAGE_AND_ORIGIN.")
```

File: llvm/docs/CMake.rst [hunk 2]
```diff
@@ -475,6 +475,17 @@ enabled sub-projects. Nearly all of these variable names begin with
 **LLVM_ENABLE_BINDINGS**:BOOL
   If disabled, do not try to build the OCaml bindings.
 
+**LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING**:STRING
```
Comment (jryans) [inline llvm/docs/CMake.rst:478]: Should the ABI-breaking-ness be mentioned here as well...? Not sure if that's a common thing to document here or not.

File: llvm/docs/CMake.rst [hunk 3]
```diff
@@ -475,6 +475,17 @@ enabled sub-projects. Nearly all of these variable names begin with
...
+  Enhances Debugify's ability to detect line number errors by storing extra
+  information inside Instructions, removing false positives from Debugify's
+  results at the cost of performance. Allowed values are `DISABLED` (default),
+  `COVERAGE`, and `COVERAGE_AND_ORIGIN`. `COVERAGE` tracks whether and why a
+  line number was intentionally dropped or not generated for an instruction,
+  allowing Debugify to avoid reporting these as errors. `COVERAGE_AND_ORIGIN`
+  additionally stores a stacktrace of the point where each DebugLoc is
```
Comment (jryans) [inline llvm/docs/CMake.rst:485]:
```suggestion
  additionally stores a stack trace of the point where each DebugLoc is
```

File: llvm/docs/CMake.rst [hunk 4]
```diff
@@ -475,6 +475,17 @@ enabled sub-projects. Nearly all of these variable names begin with
 **LLVM_ENABLE_BINDINGS**:BOOL
   If disabled, do not try to build the OCaml bindings.
 
+**LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING**:STRING
+  Enhances Debugify's ability to detect line number errors by storing extra
+  information inside Instructions, removing false positives from Debugify's
+  results at the cost of performance. Allowed values are `DISABLED` (default),
+  `COVERAGE`, and `COVERAGE_AND_ORIGIN`. `COVERAGE` tracks whether and why a
```
Comment (jryans) [inline llvm/docs/CMake.rst:482]: Should we give an example performance impact for the `COVERAGE` mode here? I think it would help people evaluate how they use the feature.

**Review by jryans — COMMENTED** [review]
This part looks good so far, just a few editorial notes. 😄

File: llvm/docs/CMake.rst (hunk 4 above)
Comment (SLTozer) [reply llvm/docs/CMake.rst:482]: Makes sense - compile time tracker reckons that [currently](http://llvm-compile-time-tracker.com/compare.php?from=9e65dcac660723a06039c7e9b30f305b9b8ca652&to=f3f169f94a0c5e9cf2711438a5d932a8d8520f60&stat=instructions:u) this sits at a slowdown of about 0.05-0.1%, depending on configuration, so I'll put something to that effect in here.

File: llvm/test/Transforms/InstCombine/reduced.ll [hunk 5]
```diff
@@ -0,0 +1,69 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+;; Tests that we do not always overwrite 
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
```
Comment (jryans) [inline llvm/test/Transforms/InstCombine/reduced.ll:3]: Hmm, did you mean to add this test here...? It seems unrelated to this PR.

Comment (SLTozer) [reply llvm/test/Transforms/InstCombine/reduced.ll:3]: I did not, my fault for `git add`ing a folder without checking my tree is clean...

**Comment by SLTozer** [comment]
@jryans Does this part look good to you now?

**Review by jryans — APPROVED** [review]
This part looks good to me. Thanks for working on this! 😄

File: llvm/cmake/modules/HandleLLVMOptions.cmake [hunk 6]
```diff
@@ -196,6 +196,16 @@ else()
   message(FATAL_ERROR "Unknown value for LLVM_ABI_BREAKING_CHECKS: \"${LLVM_ABI_BREAKING_CHECKS}\"!")
 endif()
 
+string(TOUPPER "${LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING}" uppercase_LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING)
+
+if( uppercase_LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING STREQUAL "COVERAGE" )
+  set( ENABLE_DEBUGLOC_COVERAGE_TRACKING 1 )
```
Comment (chapuni) [inline llvm/cmake/modules/HandleLLVMOptions.cmake:202]: IMO it is not a good practice to override a cached variable with a non-cached variable that has different semantics.

File: llvm/cmake/modules/HandleLLVMOptions.cmake [hunk 7]
```diff
@@ -196,6 +196,16 @@ else()
...
+elseif( uppercase_LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING STREQUAL "DISABLED" OR NOT DEFINED LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING )
+  # The DISABLED setting is default and requires no additional defines.
```
Comment (chapuni) [inline llvm/cmake/modules/HandleLLVMOptions.cmake:204]: It leaves `LLVM_ENABLE_DEBUGLOC_COVERAGE_TRACKING` as `"DISABLED"`.

File: llvm/include/llvm/Config/config.h.cmake [hunk 8]
```diff
@@ -19,6 +19,10 @@
 /* Define to 1 to enable crash memory dumps, and to 0 otherwise. */
 #cmakedefine01 LLVM_ENABLE_CRASH_DUMPS
 
+/* Define to 1 to enable expensive checks for debug location coverage checking,
+   and to 0 otherwise. */
+#cmakedefine01 ENABLE_DEBUGLOC_COVERAGE_TRACKING
```
Comment (chapuni) [inline llvm/include/llvm/Config/config.h.cmake:24]: `#cmakedefine01` handles `"DISABLED"` as `1` since it is not false-like string.

**Review by chapuni — COMMENTED** [review]
Looks like `ENABLE_DEBUGLOC_COVERAGE_TRACKING` in `llvm-config.h` is always `1`. Fixed in 1bc0b08e19788f2b34f46b183e89f5049468da2a.

**Comment by jmorse** [comment]
/me squints -- it looks like these issues were latent and then we exposed them with #143590. Thanks @chapuni for clearing this up!
